### ISSUE TITLE
CLI ignore union arguments with unsupported type

### ISSUE THREAD

**Issue by tshu-w** [omni-us/jsonargparse#168]
## 🐛 Bug report

`CLI` ignore unsupported arguments in `components` silently.

### To reproduce

```python
import jsonargparse
from ray import tune

jsonargparse.CLI(tune.Tuner)
```

```console
python tune_cli.py --help
usage: tune_cli.py [-h] [--config CONFIG] [--print_config [={comments,skip_null,skip_default}+]] [--param_space PARAM_SPACE] [--tune_config.help CLASS_PATH_OR_NAME] [--tune_config TUNE_CONFIG]
                   [--run_config.help CLASS_PATH_OR_NAME] [--run_config RUN_CONFIG]
                   {fit,restore} ...

optional arguments:
  -h, --help            Show this help message and exit.
  --config CONFIG       Path to a configuration file.
  --print_config [={comments,skip_null,skip_default}+]
                        Print the configuration after applying all other arguments and exit.

Tuner is the recommended way of launching hyperparameter tuning jobs with Ray Tune:
  --param_space PARAM_SPACE
                        Search space of the tuning job. One thing to note is that both preprocessor and dataset can be tuned here. (type: Optional[Dict[str, Any]], default: null)
  --tune_config.help CLASS_PATH_OR_NAME
                        Show the help for the given subclass of TuneConfig and exit.
  --tune_config TUNE_CONFIG
                        Tuning algorithm specific configs. Refer to ray.tune.tune_config.TuneConfig for more info. (type: Optional[TuneConfig], default: null, known subclasses: ray.tune.TuneConfig)
  --run_config.help CLASS_PATH_OR_NAME
                        Show the help for the given subclass of RunConfig and exit.
  --run_config RUN_CONFIG
                        Runtime configuration that is specific to individual trials. If passed, this will overwrite the run config passed to the Trainer, if applicable. Refer to ray.air.config.RunConfig for
                        more info. (type: Optional[RunConfig], default: null, known subclasses: ray.air.RunConfig)

subcommands:
  For more details of each subcommand add it as argument followed by --help.

  {fit,restore}
    fit                 Executes hyperparameter tuning job as configured and returns result.
    restore             Restores Tuner after a previously failed run.
```

### Expected behavior

ignoring unsupported type hints `"BaseTrainer"` in this case

https://github.com/ray-project/ray/blob/7167f9cf53eec26db9efc1a39e1914dd4cc5514c/python/ray/tune/tuner.py#L121-L128

### Environment

- jsonargparse version: 4.15.0
- Python version: 3.9.12
- How jsonargparse was installed : `pip install jsonargparse[signatures]`
- OS : macOS & Linux

**Comment by mauvilsa** [comment]
There is a warning for this, just that by default it is not printed. It is like this because these warnings are intended for the developer of the CLI, not for the end user of the CLI. To show them set the `JSONARGPARSE_DEBUG` environment variable, i.e.

```bash
$ JSONARGPARSE_DEBUG="" python tune_cli.py --help
2022-10-04 09:35:54,157	DEBUG -- signatures.py:327 -- Skipping parameter "trainable" from "Tuner.__init__" because of: Unsupported type hint typing.Union[str, typing.Callable, typing.Type[ray.tune.trainable.trainable.Trainable], ForwardRef('BaseTrainer'), NoneType].
...
```

It is a bug in the sense that according to the documentation, a `Union` type is supported as long as at least one of its subtypes is supported.

**Comment by tshu-w** [comment]
Thanks for the clarification, I changed the issue title to better match the actual bug.

**Comment by mauvilsa** [comment]
This has been fixed in commit https://github.com/omni-us/jsonargparse/commit/373d885a8dd007d0cc99968f5104e5584f6e6a03. There is a warning in debug mode which is like:
```
$ JSONARGPARSE_DEBUG="" python tune_cli.py --help
2022-10-07 12:08:09,175	DEBUG -- typehints.py:134 -- Discarding unsupported subtypes {ForwardRef('BaseTrainer')} from typing.Union[str, typing.Callable, typing.Type[ray.tune.trainable.trainable.Trainable], ForwardRef('BaseTrainer'), NoneType]
```
In the help the unsupported subtypes are not shown:
```
  --trainable TRAINABLE
                        The trainable to be tuned. (type: Union[str, Callable, Type[Trainable], null], default: null)
```

Note that even though the `tune.Tuner` class now is not ignored, I am not sure this tune cli will work. I haven't tried that class but for ray tune I ended up doing https://github.com/mauvilsa/ray-tune-cli/blob/master/ray_tune_cli.py

**Comment by tshu-w** [comment]
> This has been fixed in commit [373d885](https://github.com/omni-us/jsonargparse/commit/373d885a8dd007d0cc99968f5104e5584f6e6a03). 
Confirmed.

> Note that even though the `tune.Tuner` class now is not ignored, I am not sure this tune cli will work. 
I will try it.

> I haven't tried that class but for ray tune I ended up doing https://github.com/mauvilsa/ray-tune-cli/blob/master/ray_tune_cli.py
Yes, I know, in the past I was writing a bunch of scripts to run tune, and your example inspired me to implement a cli.

**Comment by tshu-w** [comment]
It seems like `--trainable` cannot accept `tune.Trainable` or its subclass.
```python
from jsonargparse import CLI
from ray import tune

if __name__ == "__main__":
    CLI(tune.Tuner)
```

`tune.yaml`:
```yaml
trainable:
  class_path: ray.tune.Trainable
param_space:
  seed:
    - 1234
tune_config:
  class_path: ray.tune.TuneConfig
run_config:
  class_path: ray.air.RunConfig
  init_args:
    name: blocking
    local_dir: results/ray
    log_to_file: true
```

```console
python test.py --config tune.yaml fit
usage: ray_tune_cli.py [-h] [--config CONFIG] [--print_config [={comments,skip_null,skip_default}+]] [--trainable TRAINABLE] [--param_space PARAM_SPACE] [--tune_config.help CLASS_PATH_OR_NAME]
                       [--tune_config TUNE_CONFIG] [--run_config.help CLASS_PATH_OR_NAME] [--run_config RUN_CONFIG]
                       {fit,restore} ...
ray_tune_cli.py: error: Configuration check failed :: Parser key "trainable": Value "{'class_path': 'ray.tune.Trainable'}" does not validate against any of the types in typing.Union[str, typing.Callable, typing.Type[ray.tune.trainable.trainable.Trainable], NoneType]:
  - Expected a <class 'str'> but got "{'class_path': 'ray.tune.Trainable'}"
  - Type typing.Callable expects a function or a callable class: 'ray.tune.Trainable' is not a callable class.
  - Expected a dot import path string: {'class_path': 'ray.tune.Trainable'}
  - Expected a <class 'NoneType'> but got "{'class_path': 'ray.tune.Trainable'}"
```
